### DISCUSSION TITLE
Devcontainers with remote VM: do these go together?

### DISCUSSION THREAD

**Discussion by strickvl** [devcontainers/community#38] [Q&A]
Context: I work on an open-source project where we have to support Windows, Linux and Mac operating systems as well as Python versions 3.7-3.10. It's sometimes difficult to debug issues that come up and occasionally you just have to try something out yourself to figure it out. (We run tests on the full matrix of options via Github Actions CI etc, FYI, but sometimes that doesn't quite do it).

I heard a discussion on a podcast about devcontainers and thought that maybe it's a way to address this problem. To state it clearly, I would like to be able to:

- open VS Code (either on the web directly, or on my laptop) and have that open into an environment that has all relevant dev dependencies installed
- that environment should be connected to a VM of the Python version + OS that I want (i.e. let's say Windows + Python 3.8)
- I should be able to work with the code on that environment and using branches etc as usual
- (Ideally) my VM would have some auto shutdown policy after an hour of inactivity

Do people do this already? It seems like devcontainers get me at least part of the way (the reproducible environment part), but I'm unsure how to get the VM part set up alongside it.

Any help in making this happen would be much appreciated 🙏🏻

**Comment by bamurtaugh** [comment]
Thanks for opening this discussion! And if you heard about dev containers from the Changelog podcast, that's super exciting to hear 😄.

I think this kind of setup for developing on a remote Docker host in VS Code would suit your scenario: https://code.visualstudio.com/remote/advancedcontainers/develop-remote-host. Let me know if you have additional questions or thoughts!

**Comment by borontion** [comment]
@strickvl Hi. Thanks for your sharing. I have the similar idea. The motivation for me is that I need to develop on cloud because I can access data easily (closer to data source like AWS S3, EFS, etc). I am imaging something like this:

- Use tools like [Packer](https://www.packer.io/) to define what be installed and [Terraform] to define the cloud instance spec (like instance type, disk size). This should be similar to use Dockerfile and devcontainer to define the image.
- Run the command like "open a directory on cloud", which will spin up a new instance and sync source code
- We can reuse the instance and do auto stop if there is no active window associated with.
